src/mds/MDSRank.cc

```diff
@@ -150,35 +150,19 @@ class C_Flush_Journal : public MDSInternalContext {
             << " segments to expire" << dendl;
 
     if (!expiry_gather.has_subs()) {
-      trim_segments();
+      trim_expired_segments();
       return;
     }
 
     Context *ctx = new LambdaContext([this](int r) {
-        handle_expire_segments(r);
+        ceph_assert(r == 0); // MDLog is not allowed to raise errors via
+                             // wait_for_expiry
+        trim_expired_segments();
       });
     expiry_gather.set_finisher(new MDSInternalContextWrapper(mds, ctx));
     expiry_gather.activate();
   }
 
-  void handle_expire_segments(int r) {
-    dout(20) << __func__ << ": r=" << r << dendl;
-
-    ceph_assert(r == 0); // MDLog is not allowed to raise errors via
-                         // wait_for_expiry
-    trim_segments();
-  }
-
-  void trim_segments() {
```

Comment:
IIRC, this was done this way to invoke trim_expired_segments() via finisher thread. You probably have already checked the commit history for an explanation and didn't find one? (that should be my fault) :)

Reply:
Ah, I see. Sometimes the context will complete with the `MDLog::submit_mutex` locked and cause deadlock. I'll add a note.

Reply:
PTAL